virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+TRANSMISSIVITY_COEFFICIENT = 0.50
+"""Angular coefficient of transmittivity :cite:p:`Linacre1968`"""
+FLUX_TO_ENERGY = 2.04
+"""From flux to energy conversion, umol J-1 :cite:p:`Meek1984`"""
+STEFAN_BOLTZMANN_CONSTANT = 5.67e-8
+"""Stefan-Boltzmann constant W m-2 K-4"""
+SOIL_EMISSIVITY = 0.95
+"""Soil emissivity, default for tropical rainforest"""
+CANOPY_EMISSIVITY = 0.95
+"""Canopy emissivity, default for tropical rainforest"""
+BEER_REGRESSION = 2.67e-5
+"""Parameter in equation for atmospheric transmissivity based on regression of Beer's
+radiation extinction function :cite:p:`Allen1996`"""
+CELSIUS_TO_KELVIN = 273.15
+"""Factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+
+# dummy data object, 2 grid cells, 3 canopy layers
+data: Dict[str, Any] = {
+    "elevation": np.array([100, 100], dtype=np.float32),
+    "sunshine_fraction": np.array([0.5, 1.0], dtype=np.float32),
+    "shortwave_in": np.array([100, 100], dtype=np.float32),
+    "canopy_temperature": np.array(
+        [
+            [25, 22, 20],
+            [25, 22, 20],
+        ],
+        dtype=np.float32,
+    ),
+    "surface_temperature": np.array([20, 20], dtype=np.float32),
+    "canopy_absorption": np.array(
+        [
+            [300, 200, 100],
+            [300, 200, 100],
+        ],
+        dtype=np.float32,
+    ),
+}
+
+
+@dataclass
+class Radiation:
+    """Radiation balance dataclass.
+
+    This class uses a :class:`~virtual_rainforest.core.data.Data` object to populate
+    and store radiation-related attributes which serve as inputs to other modules.
+    Elevation information from a digital elevation model is used for the topographic
+    adjustment of incoming shortwave radiation. The gross radiation that reaches the top
+    of the canopy is reduced as it penetrates through the canopy (the absorption by
+    individual canopy layers is provided by the plants module). What remains is the net
+    shortwave radiation at the surface (= forest floor), which is an input to the
+    :class:`~virtual_rainforest.models.abiotic.Energy_balance` class. Top of canopy
+    photosynthetic photon flux density is the key input for
+    :mod:`~virtual_rainforest.models.plants` on which the photosythesis is based.
+    Longwave radiation from individual canopy layers and longwave radiation from soil
+    serve as inputs to the :class:`~virtual_rainforest.models.abiotic.Energy_balance`
+    class.
+
+    The dataclass populates the elevation attribute, further attributes are
+    calculate in a __post_init__ functionality which calls a number of helper functions.
+    """
+
+    # check that elevation is above sea level
+    if np.any(data["elevation"] < 0):
+        to_raise = InitialisationError(
+            "Initial elevation contains at least one negative value!"
+        )
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    elevation: NDArray[np.float32] = data["elevation"]
+    """Elevation above sea level, [m]"""
```

Comment:
I *think* this might be the reason your tests are failing. I think this check will only be run once, when the class is defined, not every time you create a `Radiation` object. So I think all of this block needs to go into `__post_init__`, otherwise if `data["elevation"]` is later changed then a) the `elevation` field won't be updated and b) the check won't be run.

I may be wrong about this, but I think this is worth checking.

Reply:
I think that is right - in fact, that code might not run at all? I don't know how `dataclasses` handle that kind of extra code that isn't in a special method?

Reply:
Fair enough! As I said, I'd just try sticking this into `__post_init__` for now and seeing if that fixes your tests.

Reply:
we decided to change the class back to a 'normal' class and this resolved this issue :-)